### PR TITLE
feature: add check for random port, if it's used by other local services.

### PR DESCRIPTION
**Pull request by doujiang24**
1. kill nginx in clean_up when TEST_NGINX_RANDOMIZE is enable
2. add some port conflict check

### CODE REVIEW

File: lib/Test/Nginx/Util.pm [hunk 1]
```diff
@@ -578,6 +604,19 @@ sub cleanup () {
...
         undef $ChildPid;
     }
+
+    if ($Randomize) {
+        if (-d $ServRoot && $ServRoot =~ m{/t/servroot_\d+}) {
+            if (-f $PidFile) {
+                my $pid = get_pid_from_pidfile('');
+
+                if (defined $pid && $pid ne '' && is_running($pid)) {
+                    kill_nginx($pid, '');
```
Comment (agentzh) [inline lib/Test/Nginx/Util.pm]: This is duplicate since `END {}` already cleans up nginx processes. The problem here is that `clean()` gets called before the nginx process cleanup code in `END {}` so at that point in `END {}`, the pid file no longer exists. I've just fixed this in my latest `randomize` branch. Please have a look. Thanks!

File: lib/Test/Nginx/Util.pm [hunk 2]
```diff
@@ -2380,25 +2383,7 @@ END {
...
-                    last unless is_running($pid);
-
-                    sleep $TestNginxSleep;
-                    next if $i < $max_i;
-
-                    warn "WARNING: killing nginx $pid with force...";
-                    kill(SIGKILL, $pid);
-                    waitpid($pid, 0);
-                }
+                kill_nginx($pid, '');
```
Comment (agentzh) [inline lib/Test/Nginx/Util.pm]: I like this cleanup, but they'd better be in a separate PR to the master branch for their own right :)

Comment (agentzh) [inline lib/Test/Nginx/Util.pm]: @doujiang24 Never mind. I've just created the PR #55 for this myself. Please have a look. Thanks!

**Comment by doujiang24** [comment]
@agentzh  You're right, thanks for your review, I used `IO::Socket` to implement the port used check.

File: lib/Test/Nginx/Util.pm [hunk 3]
```diff
@@ -243,7 +243,28 @@ our $ForceRestartOnTest     = (defined $ENV{TEST_NGINX_FORCE_RESTART_ON_TEST})
...
+        );
+
+        last if !$sock;
+
+        $sock->close();
+
+        if ($Verbose) {
+            warn "try again, port: $ServerPort is using!";
+        }
+        $i++;
```
Comment (agentzh) [inline lib/Test/Nginx/Util.pm]: Seems like this better to use the loop form `for (my $i = 0; $i < 1000; $i++)` here?

File: lib/Test/Nginx/Util.pm [hunk 4]
```diff
@@ -243,7 +243,28 @@ our $ForceRestartOnTest     = (defined $ENV{TEST_NGINX_FORCE_RESTART_ON_TEST})
...
+        $ServerPort = int(rand 60000) + 1025;
+
+        my $sock = IO::Socket::INET->new(
+            PeerAddr => $ServerAddr,
+            PeerPort => $ServerPort,
+            Proto => 'tcp',
+            Timeout => 0.1,
+        );
+
+        last if !$sock;
```
Comment (agentzh) [inline lib/Test/Nginx/Util.pm]: `!defined $sock` is better here.

File: lib/Test/Nginx/Util.pm [hunk 5]
```diff
@@ -243,7 +243,28 @@ our $ForceRestartOnTest     = (defined $ENV{TEST_NGINX_FORCE_RESTART_ON_TEST})
...
+
+        $sock->close();
+
+        if ($Verbose) {
+            warn "try again, port: $ServerPort is using!";
```
Comment (agentzh) [inline lib/Test/Nginx/Util.pm]: `is using` should be `is in use`.

Also, `port:` should be `port`.

File: lib/Test/Nginx/Util.pm [hunk 6]
```diff
@@ -243,7 +243,28 @@ our $ForceRestartOnTest     = (defined $ENV{TEST_NGINX_FORCE_RESTART_ON_TEST})
...
+        }
+        $i++;
+    }
+
```
Comment (agentzh) [inline lib/Test/Nginx/Util.pm]: We should abort here if we do not find any ports usable instead of blindly proceed.

**Comment by doujiang24** [comment]
@agentzh All fixed :)

File: lib/Test/Nginx/Util.pm [hunk 7]
```diff
@@ -243,7 +243,30 @@ our $ForceRestartOnTest     = (defined $ENV{TEST_NGINX_FORCE_RESTART_ON_TEST})
...
+            PeerPort => $ServerPort,
+            Proto => 'tcp',
+            Timeout => 0.1,
+        );
+
+        last if !defined $sock;
+
+        $sock->close();
+
+        if ($i == 1000) {
```
Comment (agentzh) [inline lib/Test/Nginx/Util.pm]: We should avoid duplicating the magic number `1000` twice here. Better introduce a `$tries` variable to hold this value.

File: lib/Test/Nginx/Util.pm [hunk 8]
```diff
@@ -243,7 +243,30 @@ our $ForceRestartOnTest     = (defined $ENV{TEST_NGINX_FORCE_RESTART_ON_TEST})
...
+        $ServerPort = int(rand 60000) + 1025;
+
+        my $sock = IO::Socket::INET->new(
+            PeerAddr => $ServerAddr,
+            PeerPort => $ServerPort,
+            Proto => 'tcp',
+            Timeout => 0.1,
+        );
+
+        last if !defined $sock;
```
Comment (agentzh) [inline lib/Test/Nginx/Util.pm]: We should break the loop when the socket is defined instead.

File: lib/Test/Nginx/Util.pm [hunk 9]
```diff
@@ -243,7 +243,30 @@ our $ForceRestartOnTest     = (defined $ENV{TEST_NGINX_FORCE_RESTART_ON_TEST})
...
 if ($Randomize) {
     srand $$;
-    $ServerPort = int(rand 60000) + 1025;
+
+    for (my $i = 0; $i <= 1000; $i++) {
+        $ServerPort = int(rand 60000) + 1025;
+
+        my $sock = IO::Socket::INET->new(
+            PeerAddr => $ServerAddr,
+            PeerPort => $ServerPort,
```
Comment (agentzh) [inline lib/Test/Nginx/Util.pm]: Use of `PeerAddr` and `PeerPort` is wrong. They are *remote* addresses to connect to. We need a listening socket instead of a client socket.

**Comment by agentzh** [comment]
@doujiang24 Never mind, I've fixed all those issues in my local branch. Your patch does not do what it's supposed to do ;)

**Comment by agentzh** [comment]
@doujiang24 Please try the latest randomize branch on your side. Thank you!
